### DISCUSSION TITLE
bulk_save_objects method not available in async session

### DISCUSSION THREAD

**Discussion by Bechma** [sqlalchemy/sqlalchemy#6935] [Usage Questions] [answered]
Hello, I'm migrating some code from synchronous to async, and after some research of the available async session, I found out that there is no bulk_save_objects method inside. Changing to add_all make everything work fine, but now I wonder why it's not available?

**Comment by zzzeek** [comment, marked as answer]
I think they are not there probably because i've been trying to get people to stop using these methods.   if you are observing that add_all() is perfectly fine, then that would be exactly the reason why I don't like "bulk_save", because people should be using add_all().  "bulk_save" is trying to add some performance optimizations basically but we have been looking into getting these optimizations to be inhererently part of add_all().   We are not there yet for the asyncpg driver however.      if you want the absolutely highest "bulk save" performance for PostgreSQL drivers including asyncpg I would look to the example I just posted at https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#using-postgresql-on-conflict-with-returning-to-return-upserted-orm-objects .

**Reply by artkom911** [reply]
Hello. Ho offense, but you don't really work with high-load systems, do you?
it is absolutely required to have this feature in async version as it affects insert performance for such db as mysql, especially if we have remote servers. Check this out to get some info you can start with https://towardsdatascience.com/3-data-backed-ways-to-significantly-speed-up-your-mysql-bulk-inserts-28e20b42d51e

**Reply by CaselIT** [reply]
Hi,

Those operations are legacy also on the sync session as of v2, since there are better solution to achieve this.
https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-queryguide-bulk-insert

**Reply by zzzeek** [reply]
@artkom911 there are many responses to this comment.

First is, you can absolutely use the existing legacy bulk APIs with async.  Just use [session.run_sync](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#running-synchronous-methods-and-functions-under-asyncio) , which is how the asyncsession runs everything anyway:

```py
await async_session.run_sync(lambda session: session.bulk_insert_mappings(....))
```

So right off, the absence of this API feature is not like it isn't available, you can use that right now if it's what you need.

Second is, the 1.4 version of the "ORM bulk insert" methods are really not very efficient anyway and don't grant that much of a performance bump vs. regular ORM session.add(), provided in both cases the objects you provide already have their primary key values assigned.  SQLAlchemy 2.0 made a much more comprehensive change to how this all works as well so that all INSERT methods are essentially extremely fast now, relative to the 1.x series.

Third, as is illustrated in the [performance](https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.performance) examples, you can run INSERT statements directly using the ``insert()`` construct using connection.execute(), that will give you the best performance of all while still using DML (that is, SQL statements and not special APIs such as COPY).   If I have a lot of rows to insert and I'm trying to optimize the performance, that's what I'd use.  

fourth, real "bulk" INSERTs, which usually seem to be a PostgreSQL thing for people, will always be dramatically faster if you use PG COPY directly.

**Reply by CaselIT** [reply]
> SQLAlchemy 2.0 made a much more comprehensive change to how this all works as well so that all INSERT methods are essentially extremely fast now, relative to the 1.x series.

if I'm not mistaken mysql is a bit the odd one out here, since it does not support returning

**Reply by mohammad-aqajani** [reply]
```
    async with async_session() as session:
        await session.run_sync(lambda session: session.bulk_insert_mappings(MyModel, objects))
```
        
   error:
   
`   AttributeError: 'AsyncSession' object has no attribute 'bulk_insert_mappings'`

**Reply by CaselIT** [reply]
disambiguate the variable names that are use for the sync and async sessions.

**Reply by mohammad-aqajani** [reply]
Yes, I was silly! disambiguation solved the problem! 

```
    async with async_session.begin() as session:
        await session.run_sync(lambda ses: ses.bulk_insert_mappings(MyModel, objects))
        await session.commit()
```
